File: python/519.py
"""Project Euler Problem 519: Tricolored Coin Fountains.

A fountain is an arrangement of coins such that the bottom row has no gaps
and every higher coin touches two coins below it. Find the number of
3-colorings over all fountains of N coins.
"""

from __future__ import annotations

from math import isqrt
from typing import List


def solve() -> int:
    """Solve Problem 519."""
    N = 20000
    L = isqrt(2 * N)
    M = 10**9

    dp: List[List[int]] = [[0] * (L + 1) for _ in range(N + 1)]
    suf: List[List[int]] = [[0] * (L + 2) for _ in range(N + 1)]

    dp[1][1] = 3
    # Build suffix sums for n=1
    for prev in range(L, -1, -1):
        suf[1][prev] = (suf[1][prev + 1] + dp[1][prev]) % M

    for n in range(2, N + 1):
        for k in range(1, min(L, n) + 1):
            if n - k < 0:
                continue
            if k == 1:
                dp[n][k] = (2 * suf[n - 1][1]) % M
            else:
                total = suf[n - k][k - 1]
                if k == 2:
                    total = (total + dp[n - k][1]) % M
                dp[n][k] = total

        # Build suffix sums for row n
        for prev in range(L, -1, -1):
            suf[n][prev] = (suf[n][prev + 1] + dp[n][prev]) % M

    ans = sum(dp[N][1 : L + 1]) % M
    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/519.c
/*
 * Project Euler Problem 519: Tricolored Coin Fountains.
 * Count 3-colorings over all fountains of N coins.
 * DP with suffix sums.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef long long ll;

#define N_VAL 20000
#define M_VAL 1000000000LL

int main() {
    int N = N_VAL;
    int L = (int)sqrt(2.0 * N) + 1;
    ll M = M_VAL;

    /* dp[n][k] and suf[n][k] - use rolling arrays since we only need dp[n-k] */
    /* Actually, dp[n][k] depends on suf[n-k][k-1] and dp[n-k][1], which means
       we need access to all previous rows. Allocate full arrays. */

    /* Use flattened arrays: dp[n*(L+1) + k] */
    ll *dp = (ll*)calloc((ll)(N + 1) * (L + 1), sizeof(ll));
    ll *suf = (ll*)calloc((ll)(N + 1) * (L + 2), sizeof(ll));

    #define DP(n, k) dp[(ll)(n) * (L + 1) + (k)]
    #define SUF(n, k) suf[(ll)(n) * (L + 2) + (k)]

    DP(1, 1) = 3;
    for (int prev = L; prev >= 0; prev--)
        SUF(1, prev) = (SUF(1, prev + 1) + DP(1, prev)) % M;

    for (int n = 2; n <= N; n++) {
        int kmax = (L < n) ? L : n;
        for (int k = 1; k <= kmax; k++) {
            if (n - k < 0) continue;
            if (k == 1) {
                DP(n, k) = (2 * SUF(n - 1, 1)) % M;
            } else {
                ll total = SUF(n - k, k - 1);
                if (k == 2)
                    total = (total + DP(n - k, 1)) % M;
                DP(n, k) = total % M;
            }
        }
        for (int prev = L; prev >= 0; prev--)
            SUF(n, prev) = (SUF(n, prev + 1) + DP(n, prev)) % M;
    }

    ll ans = 0;
    for (int k = 1; k <= L; k++)
        ans = (ans + DP(N, k)) % M;

    printf("%lld\n", ans);

    free(dp);
    free(suf);
    return 0;
}
